### PR TITLE
[Feature Branch] Yolov8

### PR DESCRIPTION
**Pull request by corey-nm**
Adds yolov8 training script to the repo.

# Test plan

## Recipe

```yaml
version: 1.1.0

training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0
    end_epoch: 15

pruning_modifiers:
  - !GMPruningModifier
    init_sparsity: 0.05
    final_sparsity: 0.10
    start_epoch: 5.0
    end_epoch: 10.0
    update_frequency: 1.0
    params: ["model.0.conv.weight", "model.1.conv.weight"]

quantization_modifiers:
  - !QuantizationModifier
    start_epoch: 11.0
    freeze_bn_stats_epoch: 12.0
    disable_quantization_observer_epoch: 13.0
    ignore: ["Upsample", "Concat", "SiLU"]
```

There are two parts to this:
1. the DFL final layer in the yolo models are detached (they do not require grads). Therefore, the pruner must ignore them as this currently crashes (see potential fix in #1329).
2. Due to ultralytics adding special attributes to top layer modules of the model, we must ignore Upsample & Concat for quantization. These attributes are [set on this line](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/tasks.py#L451), and [used on this line](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/tasks.py#L53). Without ignoring Upsample & Concat, these two layers get replaced with `QuantWrapper`, which does not have these special attributes set.

### Base ✅

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --name dense-base`

### From local yolo checkpoint ✅

After running the base training command above:

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --name dense-local-yolo --model runs/detect/dense-base/weights/best.pt`

Should properly print:
```
...
Transferred 355/355 items from pretrained weights
...
```

### Continuing an interrupted run ✅

After running the base training command above:
`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --name dense-local-yolo --resume --model runs/detect/dense-base/weights/last.pt`

Properly prints:
```
...
Loading local ultralytics checkpoint
...
Transferred 355/355 items from pretrained weights
...
Resuming training from <path>/weights/last.pt from epoch 5 to 15 total epochs
```

### Base ✅

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --name pq --recipe yolov8-pq.yaml`

- Should train for 15 epochs.
- Pruning should be logged & change in wandb
- QuantizationModifier should be logged & wandb after it is applied

### Continuing from an interrupted run before quantization ✅

Run training for a few epochs, ctrl+c, then use the following command:

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --resume --model <path>/weights/last.pt`

You should see the following message:

```
...
Loading local sparseml checkpoint
...
Applying structure from un-finished recipe in checkpoint at epoch <>
Loaded previous weights from sparseml checkpoint
...
```

### Continuing from an interrupted run after quantization ✅

Run training for a few epochs until quantization is applied, ctrl+c, then use the following command:

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --resume --model <path>/weights/last.pt`

You should see:
```
...
Loading local sparseml checkpoint
...
Applying structure from un-finished recipe in checkpoint at epoch <>
...
```

### **NOT SUPPORTED** Fine tuning a quantized model with no recipe provided ❌ 

Run training with recipe to completion, then run:

`CUDA_VISIBLE_DEVICES=0 python src/sparseml/pytorch/yolov8/train.py --model <path>/weights/last.pt`

You should see:
```
...
Loading local sparseml checkpoint
...
Applying structure from completed recipe in checkpoint at epoch <>
...
```

### CODE REVIEW

**Comment by KSGulin** [comment]
Should we move this to sparseml/yolov8, on the same directory level as yolov5? For context, yolov5 was placed there for high visibility, as it is it's own thing like transformers is, rather than a native pytorch implementation

**Comment by KSGulin** [comment]
@corey-nm @bfineran is there a plan to integrate into the `ultralytics/ultralytics` upstream as we are in yolov5? If so, may be good to keep that in mind with this integration. Since we're going for a forkless integration, would the target then be to have two separate integrations? One in sparseml and one in the ultralytics upstream?

**Comment by corey-nm** [comment]
> Should we move this to sparseml/yolov8, on the same directory level as yolov5? For context, yolov5 was placed there for high visibility, as it is it's own thing like transformers is, rather than a native pytorch implementation

I would say high visibility will be achieved through documentation (README/sample code/etc). I don't expect people will be discovering we support yolov8 through looking at the src directory, and this way we know it depends on pytorch.

**Comment by corey-nm** [comment]
> is there a plan to integrate into the `ultralytics/ultralytics` upstream as we are in yolov5? If so, may be good to keep that in mind with this integration. Since we're going for a forkless integration, would the target then be to have two separate integrations? One in sparseml and one in the ultralytics upstream?

If we get it into upstream I would vote for removing our sparseml version of it. I haven't heard plans of contributing to upstream though?

**Comment by KSGulin** [comment]
> I would say high visibility will be achieved through documentation (README/sample code/etc). I don't expect people will be discovering we support yolov8 through looking at the src directory, and this way we know it depends on pytorch.

Fair enough, I think the argument could be made either way and I can't say I feel too strongly about either option. But I do feel that it's counterintuitive to have yolov5 and yolov8 in different directories

**Comment by KSGulin** [comment]
> If we get it into upstream I would vote for removing our sparseml version of it. I haven't heard plans of contributing to upstream though?

My understanding is that an upstream push will be a part of the yolov8 integration. Probably worth further discussion with the team

**Comment by KSGulin** [comment]
Nice workaround for the detached layers. To complete testing coverage I'd recommend adding:

- Continuing an interrupted run with the --resume keyword (base)
- Continuing an interrupted run with the --resume keyword (recipe)
- Continuing fine-tuning quantized model without recipe

File: src/sparseml/pytorch/yolov8/trainers.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+class SparseDetectionTrainer(SparseTrainer, DetectionTrainer):
+    pass
+
+
+class SparseClassificationTrainer(SparseTrainer, ClassificationTrainer):
+    pass
+
+
+class SparseSegmentationTrainer(SparseTrainer, SegmentationTrainer):
```
Comment (bfineran) [inline src/sparseml/pytorch/yolov8/trainers.py:398]: nice

**Review by bfineran — APPROVED** [review]
LGTM @corey-nm! As a quick follow up - let's get the PR description for this adapted into a rough README (to be iterated on by product/research)

**Review by rahul-tuli — APPROVED** [review]

**Pull request merged**